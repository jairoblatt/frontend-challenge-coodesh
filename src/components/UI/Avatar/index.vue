<template>
  <component :is="tag" class="avatar" :style="styles">
    <slot />
  </component>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Avatar",

  props: {
    tag: {
      type: String,
      default: "div",
    } as PropOptions<string>,
    // The prop size it's definedly with 'rem' css unit meansure
    size: {
      type: [String, Number],
      default: 2,
    } as PropOptions<string | number>,
    // The prop border it's definedly with 'px' css unit meansure
    border: [String, Number] as PropType<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        height: `${this.size}rem`,
        width: `${this.size}rem`,
        borderWidth: `${this.border || Number(this.size) / 2}px`,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.avatar {
  @apply rounded-full border-4 border-primary-base border-solid overflow-hidden;
}

.avatar img,
source {
  @apply min-w-full h-auto object-contain;
}

/* absolute -top-28 left-2/4 transform -translate-x-2/4 */

/* .user__avatar img,
source {
  height: 12rem;
  width: 12rem;
} */
</style>
