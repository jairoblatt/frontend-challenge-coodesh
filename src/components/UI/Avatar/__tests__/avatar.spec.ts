import { mount, Wrapper, MountOptions } from "@vue/test-utils";
import Avatar from "../index.vue";

describe("Avatar.vue", () => {
  type Instance = InstanceType<typeof Avatar>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Avatar, {
        ...options,
      });
    };
  });

  it("Should mount component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should be render the current element when the prop tag is passed", () => {
    const wrapper = mountFunction({
      propsData: {
        tag: "small",
      },
    });

    expect(wrapper.find("small")).toBeDefined();
  });

  it("Should apply the styles size and border when the respective props is passed", () => {
    const wrapper = mountFunction({
      propsData: {
        size: 4,
        border: 10,
      },
    });

    const divElement = wrapper.find("div");

    expect(divElement.attributes("style")).toBe("height: 4rem; width: 4rem; border-width: 10px;");
  });

  it("Should render the slot content when them it is passed", () => {
    const slotContent = `<h2>Something</h2>`;

    const wrapper = mountFunction({
      slots: {
        default: slotContent,
      },
    });
    expect(wrapper.find("h2").text()).toBe("Something");
  });
});
