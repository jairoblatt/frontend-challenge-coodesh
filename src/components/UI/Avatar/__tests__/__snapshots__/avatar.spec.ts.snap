// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar.vue Should mount component 1`] = `"<div class=\\"avatar\\" style=\\"height: 2rem; width: 2rem; border-width: 1px;\\"></div>"`;
